stellargraph/layer/rgcn.py

```diff
@@ -0,0 +1,554 @@
+    should be supplied using the :class:`RelationalFullBatchNodeGenerator` class.
+    The generator object should be instantiated as follows::
+
+        generator = RelationalFullBatchNodeGenerator(G)
+
+    Note that currently the RGCN class is compatible with both sparse and dense adjacency
+    matrices and the :class:`RelationalFullBatchNodeGenerator` will default to sparse.
+
+    For more details, please see the RGCN demo notebook:
+
+    emos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb
+
+    Notes:
+      - The inputs are tensors with a batch dimension of 1. These are provided by the \
+        :class:`RelationalFullBatchNodeGenerator` object.
+
+      - The nodes provided to the :class:`RelationalFullBatchNodeGenerator.flow` method are
+        used by the final layer to select the predictions for those nodes in order.
+        However, the intermediate layers before the final layer order the nodes
+        in the same way as the adjacency matrix.
+
+    Examples:
+        Creating a RGCN node classification model from an existing :class:`StellarGraphBase`
+        object ``G``::
+
+            generator = RelationalFullBatchNodeGenerator(G)
+            rgcn = RGCN(
+                    layer_sizes=[32, 4],
+                    activations=["elu","softmax"],
+                    bases=10,
+                    generator=generator,
+                    dropout=0.5
+                )
+            x_inp, predictions = rgcn.node_model()
+
+    Args:
+        layer_sizes (list of int): Output sizes of RGCN layers in the stack.
+        generator (RelationalFullBatchNodeGenerator): The generator instance.
+        num_bases (int): Specifies number of basis matrices to use for the weight matrics of the RGCN layer
+            as in the paper. Defaults to 0 which specifies that no basis decomposition is used.
+        bias (bool): If True, a bias vector is learnt for each layer in the RGCN model.
+        dropout (float): Dropout rate applied to input features of each RGCN layer.
+        activations (list of str or func): Activations applied to each layer's output;
+            defaults to ['relu', ..., 'relu'].
+        kernel_regularizer (str or func): The regulariser to use for the weights of each layer;
+            defaults to None.
+    """
+
+    def __init__(
+        self,
+        layer_sizes,
+        generator,
+        bias=True,
+        num_bases=0,
+        dropout=0.0,
+        activations=None,
+        **kwargs
+    ):
+        if not isinstance(generator, RelationalFullBatchNodeGenerator):
+            raise TypeError(
+                "Generator should be a instance of RelationalFullBatchNodeGenerator"
+            )
+
+        n_layers = len(layer_sizes)
+        self.layer_sizes = layer_sizes
+        self.activations = activations
+        self.bias = bias
+        self.num_bases = num_bases
+        self.dropout = dropout
+        self.generator = generator
```

Comment:
I'm trying to move away from storing the generator in the model. It seems too messy.

I have also introduced a 'multiplicity' property to the generators which is 1 for a node inference and 2 for link inference. To match the new-style full batch generators you could look at that.

See this PR: https://github.com/stellargraph/stellargraph/pull/543

Reply:
yeah agreed, your new method is way nicer - done!